src/bundle.c

```diff
@@ -1058,13 +1058,190 @@ static struct list *get_dir_files_sorted(char *path)
+	struct list *iter = alias_definitions;
+
+	while (iter) {
+		struct alias_lookup *lookup = (struct alias_lookup *)iter->data;
+		iter = iter->next;
+		if (strcmp(lookup->alias, alias) == 0) {
+			bundles = list_deep_clone_strs(lookup->bundles);
+			break;
+		}
+	}
+	if (!bundles) {
+		bundles = list_prepend_data(bundles, strdup_or_die(alias));
+	}
+
+	return bundles;
+}
+
+/* Parse an alias definition file for a list of bundles
+ * it specifies
+ *
+ * Example file:
+ * alias1	b1	b2
+ * alias2	b3	b4
+ */
+static struct list *parse_alias_file(char *fullpath)
+{
+	FILE *afile = NULL;
+	struct list *alias_content = NULL;
+	char *line = NULL;
+	size_t len = 0;
+	ssize_t read = -1;
+
+	afile = fopen(fullpath, "r");
+	if (!afile) {
+		return NULL;
+	}
+
+	while ((read = getline(&line, &len, afile)) != -1) {
+		struct alias_lookup *l = NULL;
+		char *c = NULL;
+
+		/* get the alias but ignore lines without bundles */
+		c = strchr(line, '\t');
+		if (!c) {
+			continue;
+		}
+
+		/* strip the newline if it exists */
+		if (line[read - 1] == '\n') {
+			line[read - 1] = 0;
+		}
+
+		*c = 0;
+		c++;
+
+		l = calloc(1, sizeof(struct alias_lookup));
+		ON_NULL_ABORT(l);
+
+		/* get the bundles associated with the alias
+		 * note a file with multiple copies of the same
+		 * alias will be first alias defintion wins */
+		while (c && *c) {
+			char *next = strchr(c, '\t');
+			char *bundle = NULL;
+			/* two tabs in a row, skip */
+			if (c == next) {
+				c++;
+				continue;
+			}
+			if (next) {
+				*next = 0;
+				next++;
+			}
+			bundle = strdup_or_die(c);
+			l->bundles = list_prepend_data(l->bundles, bundle);
+			c = next;
+		}
+		if (!l->bundles) {
+			free(l);
+			continue;
+		}
+		l->alias = strdup_or_die(line);
+		alias_content = list_prepend_data(alias_content, l);
+	}
+
+	fclose(afile);
+	if (line) {
+		free(line);
+	}
+	return alias_content;
+}
+
+/* Get alias definitions using both system and user
+ * alias definition files
+ */
+static struct list *get_alias_definitions()
+{
+	char *path = NULL;
+	struct list *combined = NULL;
+	struct list *definitions = NULL;
+	struct list *iters = NULL;
+	struct list *iteru = NULL;
+	struct list *system_alias_files = NULL;
+	struct list *user_alias_files = NULL;
+
+	/* get sorted system and user filenames */
+	string_or_die(&path, "%s/%s", path_prefix, SYSTEM_ALIAS_PATH);
+	system_alias_files = get_dir_files_sorted(path);
+	free(path);
+
+	string_or_die(&path, "%s/%s", path_prefix, USER_ALIAS_PATH);
+	user_alias_files = get_dir_files_sorted(path);
+	free(path);
+
+	/* get a combined list with user files overriding system files */
+	iters = system_alias_files;
+	iteru = user_alias_files;
+	while (iters && iteru) {
+		int pivot = strcmp(basename(iteru->data), basename(iters->data));
+		if (pivot == 0) {
+			if (iters == system_alias_files) {
+				system_alias_files = iters->next;
+			}
+			iters = list_free_item(iters, free);
+
+			combined = list_prepend_data(combined, iteru->data);
```

Comment:
this changes the order of user config files. If you have:

    /etc/swupd/alias.d/a1
    /etc/swupd/alias.d/b1
    /usr/share/defaults/swupd/alias.d/b1

We will look for aliases in the order:

    /etc/swupd/alias.d/b1
    /etc/swupd/alias.d/a1
    /usr/share/defaults/swupd/alias.d/b1

You can just remove the item from iters and concat system_alias_files with user_alias_files later.

Reply:
Ah great catch, added test for this and will fix.